state/storagePruningManager/evictionWaitingList/memoryEvictionWaitingList.go

```diff
@@ -0,0 +1,167 @@
+package evictionWaitingList
+
+import (
+	"bytes"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go-core/core/check"
+	"github.com/ElrondNetwork/elrond-go-core/data"
+	"github.com/ElrondNetwork/elrond-go-core/marshal"
+	"github.com/ElrondNetwork/elrond-go/state"
+	"github.com/ElrondNetwork/elrond-go/state/temporary"
+)
+
+type hashInfo struct {
+	roothashes [][]byte
+}
+
+// memoryEvictionWaitingList is a structure that caches keys that need to be removed from a certain database.
+// If the cache is full, the caches will be emptied automatically. Writing at the same key in
+// cacher and db will overwrite the previous values.
+type memoryEvictionWaitingList struct {
+	cache         map[string]temporary.ModifiedHashes
+	reversedCache map[string]*hashInfo
+	cacheSize     uint
+	marshalizer   marshal.Marshalizer
+	opMutex       sync.RWMutex
+}
+
+// NewMemoryEvictionWaitingListV2 creates a new instance of memoryEvictionWaitingList
+func NewMemoryEvictionWaitingListV2(size uint, marshalizer marshal.Marshalizer) (*memoryEvictionWaitingList, error) {
+	if size < 1 {
+		return nil, data.ErrInvalidCacheSize
+	}
+	if check.IfNil(marshalizer) {
+		return nil, data.ErrNilMarshalizer
+	}
+
+	return &memoryEvictionWaitingList{
+		cache:         make(map[string]temporary.ModifiedHashes),
+		reversedCache: make(map[string]*hashInfo),
+		cacheSize:     size,
+		marshalizer:   marshalizer,
+	}, nil
+}
+
+// Put stores the given hashes in the eviction waiting list, in the position given by the root hash
+func (mewl *memoryEvictionWaitingList) Put(rootHash []byte, hashes temporary.ModifiedHashes) error {
+	mewl.opMutex.Lock()
+	defer mewl.opMutex.Unlock()
+
+	log.Trace("trie eviction waiting list", "size", len(mewl.cache))
+
+	mewl.putInReversedCache(rootHash, hashes)
+
+	if uint(len(mewl.cache)) <= mewl.cacheSize {
+		mewl.cache[string(rootHash)] = hashes
+		return nil
+	}
+
+	log.Warn("trie nodes eviction waiting list full, emptying...")
+	mewl.cache = make(map[string]temporary.ModifiedHashes)
```

Comment:
why do we clear the entire cache on Put?

Reply:
As agreed, when the pruning buffer is full, we should empty it in order to be able to process the future pruning hashes.